### PR TITLE
feat(ws): hold off ws clean below a mining threshold

### PR DESCRIPTION
**Pull request by agent-refr**
> **AI-assisted change proposal.** Filed by agent driven by @Cervator via [GDD](https://siliconsaga.github.io/yggdrasil/gdd/).

## Summary

`ws clean` now embeds a mining-cadence check instead of deleting unconditionally. The scratch files it sweeps (`.commits/`, `.crs/`, `.issues/`, `.outputs/`, `.tmp/`) double as mining material for the gdd-housekeeping "mine scratch dirs" step — clearing them reflexively after every commit/CR starves that step. The threshold puts the cadence in the tool itself, so it doesn't depend on the agent remembering a convention.

- **Count-first, then decide.** `ws clean` tallies what it would remove. If the total is **below `$WS_CLEAN_MINE_THRESHOLD` (default 50)** and `--force` is not given, it prints a per-dir breakdown + a "holding off" nudge and deletes nothing. At/above threshold, or with `--force`, it cleans as before.
- **`--force` override.** The agent re-runs with `--force` when the user explicitly asked to clean, or right after a mining pass — otherwise it backs off. Env-overridable floor via `WS_CLEAN_MINE_THRESHOLD` (invalid values warn + fall back to 50).
- **Dispatcher `ROOT_DIR` honoring.** Changed the `ws` dispatcher to honor a pre-set `ROOT_DIR` (`: "${ROOT_DIR:=…}"`, matching `ws-commit.sh`) so the new tests can isolate against a temp workspace. Unset in normal use → still resolves to the script's parent, so production behavior is unchanged.

## Test plan

- [x] `bash scripts/ws test yggdrasil` — full suite green (**182** tests, +9 new; 2 platform-skipped on Windows).
- [x] New `tests/ws-clean/` suite: threshold boundary (below holds off, at/above cleans), `--force` override, env threshold + invalid-value fallback, multi-dir + `.tmp` counting, `--help`, unknown-flag error.
- [x] Manual: isolated-dir runs confirm hold-off (3 files / threshold 5) and clean (2 files / threshold 2, and `--force`).

## Notes for reviewers

- The `ROOT_DIR` change is the one bit beyond `ws clean` itself — it's a behavior-preserving consistency fix (the dispatcher was the only thing overwriting an env-supplied `ROOT_DIR`; `ws-commit.sh` already used `:=`). It's what makes the scratch-dir tests isolate instead of operating on the real workspace.
- Default threshold 50 is a starting guess; it's env-tunable, so we can adjust after seeing real accumulation rates.

## Summary by CodeRabbit

* **New Features**
  * CLI: `clean` command now supports `--force` and `--help`, performs a two-phase scan of draft and temporary locations, prints a per-location breakdown, respects a configurable threshold (holds off with guidance unless forced), deletes eligible drafts and top-level temp entries while preserving the temp container, and reports a final cleaned-count message.

* **Tests**
  * Added comprehensive tests for threshold gating, `--force`, `--help`, counting across locations, edge cases (spaces, empty scratch), and invalid-threshold handling.

[Review Change Stack](https://app.coderabbit.ai/change-stack/SiliconSaga/yggdrasil/pull/73?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: scripts/ws [hunk 1]
```diff
@@ -286,57 +289,121 @@ HELP
...
+    # .tmp/ contents are arbitrary and often nested (synthetic git repos,
+    # test fixtures, generated scripts), so recursive cleanup. We blow
+    # away top-level entries; the directory itself stays for next-use.
+    if [[ -d "$tmp_dir" && "$tmp_count" -gt 0 ]]; then
+        find "$tmp_dir" -mindepth 1 -maxdepth 1 -exec rm -rf {} +
+        echo "  .tmp/  $tmp_count entr(ies) removed"
+        removed=$((removed + tmp_count))
+    fi
+
+    echo "Cleaned $removed draft file(s) total."
```
[the remark this answers is not in the document]
Comment (agent-refr) [reply scripts/ws]: Fixed in 4082695 — ws clean now collects the exact paths in a single scan and deletes that snapshot, so the threshold check and the deletion act on the same set. The re-scan is gone, which also removes the duplicated find. Thanks for catching the TOCTOU window.

File: tests/ws-clean/test_helper.bash [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+if command -v timeout >/dev/null 2>&1; then
+    TIMEOUT_BIN="timeout"
+elif command -v gtimeout >/dev/null 2>&1; then
+    TIMEOUT_BIN="gtimeout"
+else
+    echo "ERROR: neither 'timeout' nor 'gtimeout' found on PATH." >&2
+    echo "  Install GNU coreutils — on macOS: 'brew install coreutils'." >&2
+    echo "  See tests/README.md." >&2
+    exit 1
+fi
```
[the remark this answers is not in the document]
Comment (agent-refr) [reply tests/ws-clean/test_helper.bash]: Fixed in 4082695 — moved detection into a detect_timeout_bin function called from setup via init_clean_workspace, using bats skip. A missing timeout/gtimeout now skips the affected tests with a clear message instead of aborting file load. Noted: the ws-smoke helper has the same source-time pattern, left out of this PR scope as a possible follow-up.

File: scripts/ws [hunk 3]
```diff
@@ -286,57 +289,121 @@ HELP
...
-        echo "Usage: ws clean" >&2
-        echo "" >&2
-        echo "Remove draft files from .issues/, .crs/, .commits/, .outputs/," >&2
-        echo "and the workspace scratch area .tmp/." >&2
-        exit 1
+    local force=0
+    local arg
+    for arg in "$@"; do
+        case "$arg" in
+            --force|--yes) force=1 ;;
```
[the remark this answers is not in the document]
Comment (agent-refr) [reply scripts/ws]: Fixed in 4082695 — dropped the --yes alias. force is now the single documented and tested override.

File: scripts/ws [hunk 4]
```diff
@@ -286,57 +289,113 @@ HELP
...
+    # generated scripts), so recursive removal; the .tmp/ dir itself stays
+    # for next-use since we only remove its top-level entries.
+    printf '%s\n' "${report[@]}"
+    local c
+    for c in ${candidates[@]+"${candidates[@]}"}; do
+        rm -f "$c"
+    done
+    local t
+    for t in ${tmp_entries[@]+"${tmp_entries[@]}"}; do
+        rm -rf "$t"
```
[the remark this answers is not in the document]
Comment (agent-refr) [reply scripts/ws]: Fixed in bfcffd8 — good catch that the guard form still splits. Replaced both loops with a single length-guarded, quoted, batched rm -f -- "${candidates[@]}" / rm -rf -- "${tmp_entries[@]}". The length guard keeps the empty-array safety on older bash. Added a regression test (force clean handles paths with spaces) that the old loop failed.

File: scripts/ws [hunk 5]
```diff
@@ -286,57 +289,113 @@ HELP
...
+    done
```
[the remark this answers is not in the document]
Comment (agent-refr) [reply scripts/ws]: Fixed in bfcffd8 — batched into a single rm -f -- "${candidates[@]}" (guarded for non-empty), so no more per-file process spawn.

[the remark this answers is not in the document]
Comment (agent-refr) [reply scripts/ws]: Fixed in bfcffd8 — same batching for the .tmp entries: a single guarded rm -rf -- "${tmp_entries[@]}" instead of one rm per entry.

**Review by Cervator — APPROVED** [review]
